File: JS - Fundamentals/13. Exam Prep 2/03. P!rates/pirates.js
function solve(arr){
    const cities = {};
    const actions = {
        Plunder : (cities, [name, killed, stolen])=>{
            killed = Number(killed);
            stolen = Number(stolen);

            cities[name].population -= killed;
            cities[name].gold -= stolen;
            console.log(`${name} plundered! ${stolen} gold stolen, ${killed} citizens killed.`);
            
            if(cities[name].population <= 0 || cities[name].gold <= 0){
                delete cities[name];
                console.log(`${name} has been wiped off the map!`);
            }
        },
        Prosper : (cities, [name, addedGold]) =>{
            addedGold = Number(addedGold);

            if(addedGold < 0){
                console.log(`Gold added cannot be a negative number!`);
            } else {
                cities[name].gold += addedGold;

                console.log(`${addedGold} gold added to the city treasury. ${name} now has ${cities[name].gold} gold.`);
            }
        }
    }

    while((line = arr.shift()) !== 'Sail'){
        let [name, population, gold] = line.split('||');
        population = Number(population);
        gold = Number(gold);

        if(!cities.hasOwnProperty(name)){
            cities[name] = {population, gold};
        } else {
            cities[name].population += population;
            cities[name].gold += gold;
        }
    }

    while((line = arr.shift()) !== 'End'){
        const [command, ...params] = line.split('=>');
        const action = actions[command];
        action(cities, params);
    }

    if(Object.entries(cities).length > 0){
        console.log(`Ahoy, Captain! There are ${Object.entries(cities).length} wealthy settlements to go to:`);

        Object
        .entries(cities)
        .sort(sortTowns)
        .forEach(town => {
            console.log(`${town[0]} -> Population: ${town[1].population} citizens, Gold: ${town[1].gold} kg`);
        })
    } else {
        console.log('Ahoy, Captain! All targets have been plundered and destroyed!');
    }


    function sortTowns(a, b){
        return b[1].gold - a[1].gold || a[0].localeCompare(b[0])
    }
}

solve([
    'Tortuga||345000||1250',
    'Santo Domingo||240000||630',
    'Havana||410000||1100',
    'Sail',
    'Plunder=>Tortuga=>75000=>380',
    'Prosper=>Santo Domingo=>180',
    'End'
  ])

File: Java Fundamentals/11. Exam Prep 1/src/Pirates.java
import java.util.*;

public class Pirates {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);

        Map<String, List<Integer>> towns = new LinkedHashMap<>();
        String input = scan.nextLine();

        input = setTowns(input, towns, scan);

        input = scan.nextLine();

        townActions(input, towns, scan);

        if(towns.entrySet().size() > 0){
            System.out.printf("Ahoy, Captain! There are %d wealthy settlements to go to:\n", towns.entrySet().size());

            towns
                    .entrySet()
                    .stream()
                    .sorted(Map.Entry.comparingByKey())
                    .sorted((a,b) -> b.getValue().get(1) - a.getValue().get(1))
                    .forEach(town -> System.out.printf("%s -> Population: %d citizens, Gold: %d kg\n", town.getKey(), town.getValue().get(0), town.getValue().get(1)));
        } else {
            System.out.println("Ahoy, Captain! All targets have been plundered and destroyed!");
        }
    }

    public static String setTowns(String input, Map<String, List<Integer>> towns, Scanner scan){
        while(!input.equals("Sail")) {
            String[] inputData = input.split("\\|\\|");
            String townName = inputData[0];
            int population = Integer.parseInt(inputData[1]);
            int gold = Integer.parseInt(inputData[2]);

            towns.putIfAbsent(townName, new ArrayList<>());
            List<Integer> currentCityData = towns.get(townName);

            if (currentCityData.size() > 0) {
                currentCityData.set(0, currentCityData.get(0) + population);
                currentCityData.set(1, currentCityData.get(1) + gold);
            } else {
                currentCityData.add(population);
                currentCityData.add(gold);
            }

            input = scan.nextLine();
        }
        return input;
    }

    public static void townActions(String input, Map<String, List<Integer>> towns, Scanner scan){
        while(!input.equals("End")){
            String[] data = input.split("=>");
            String command = data[0];
            String townName = data[1];
            List<Integer> currentTownData = towns.get(townName);

            if(command.equals("Plunder")){
                int killedPeople = Integer.parseInt(data[2]);
                int stolenGold = Integer.parseInt(data[3]);


                int currentPopulation = currentTownData.get(0);
                int currentGold = currentTownData.get(1);

                currentTownData.set(0, currentPopulation - killedPeople);
                currentTownData.set(1, currentGold - stolenGold);

                System.out.printf("%s plundered! %d gold stolen, %d citizens killed.\n", townName, stolenGold, killedPeople);

                if(currentTownData.get(0) <= 0 || currentTownData.get(1) <=0){
                    towns.remove(townName);
                    System.out.printf("%s has been wiped off the map!\n", townName);
                }
            } else if(command.equals("Prosper")){
                int addedGold = Integer.parseInt(data[2]);

                if(addedGold <= 0){
                    System.out.println("Gold added cannot be a negative number!");
                } else {
                    currentTownData.set(1, currentTownData.get(1) + addedGold);
                    System.out.printf("%d gold added to the city treasury. %s now has %d gold.\n", addedGold, townName, currentTownData.get(1));
                }
            }

            input = scan.nextLine();
        }
    }
}
